File: benes/benes.py
#!/usr/bin/python
#
# arbitrary-sized Waksman network experimental impl
# (C) 2020 Riad S. Wahby

import sys
if sys.version_info[0] != 3:
    raise RuntimeError("python3 required")

class BenesNetworkBase(object):
    num_inputs = 0
    ord_out = None
    sw = None

    def __init__(self, num_inputs):
        self.num_inputs = num_inputs

    def reset(self):
        self.ord_out = None
        self.sw = None

    def _set_order(self, ord_in, ord_out):
        assert len(ord_in) == self.num_inputs   # ord_in has right number of values
        assert len(ord_in) == len(set(ord_in))  # ord_in values are unique
        assert set(ord_in) == set(ord_out)      # ord_out is permutation of ord_in
        # rewrite ord_out in terms of input indices
        # this lets us pass lists of arbitrary values for ord_in and ord_out
        reord = dict(zip(ord_in, range(0, len(ord_in))))
        self.ord_out = [ reord[v] for v in ord_out ]

    def set_switches(self, ord_in, ord_out):
        assert self.sw is None, "set_switches called twice"
        self._set_order(ord_in, ord_out)

    def route(self, inputs):
        assert self.sw is not None
        assert len(inputs) == self.num_inputs
        return [ inputs[idx] for idx in self.ord_out ]

class BenesNetwork2x2(BenesNetworkBase):
    def __init__(self):
        super().__init__(2)

    def set_switches(self, ord_in, ord_out):
        super().set_switches(ord_in, ord_out)
        self.sw = self.ord_out[0] != 0

    def route(self, inputs):
        r_exp = super().route(inputs)
        if self.sw:
            r = [inputs[1], inputs[0]]
        else:
            r = list(inputs)
        assert r_exp == r
        return r

class BenesNetwork3x3(BenesNetworkBase):
    def __init__(self):
        super().__init__(3)

    #              t0
    # --|---|-------------|---|--
    #   | I |             | O |
    # --|---|--+       +--|---|--
    #          |       |
    #        m +-|---|-+ t1
    #            | M |
    # -----------|---|-----------
    def set_switches(self, ord_in, ord_out):
        super().set_switches(ord_in, ord_out)

        if self.ord_out[2] != 2:
            sw_m = True
            sw_i = self.ord_out[2] == 0
            sw_o = self.ord_out[1] != 2
        else:
            sw_m = False
            sw_i = False
            sw_o = self.ord_out[0] == 1

        self.sw = [sw_i, sw_m, sw_o]

    def route(self, inputs):
        r_exp = super().route(inputs)
        r = [None] * 3
        (sw_i, sw_m, sw_o) = self.sw

        (t0, m) = inputs[1::-1] if sw_i else inputs[:2]
        (t1, r[2]) = (inputs[2], m) if sw_m else (m, inputs[2])
        (r[0], r[1]) = (t1, t0) if sw_o else (t0, t1)
        assert None not in r
        assert r_exp == r
        return r

class BenesNetwork(BenesNetworkBase):
    top = None
    bot = None
    sw = None

    def __init__(self, num_inputs):
        super().__init__(num_inputs)
        assert num_inputs >= 4, "BenesNetwork requires at least 4 inputs"
        size_top = num_inputs // 2
        size_bot = num_inputs - size_top
        insts = [None, None]
        for (idx, size) in enumerate((num_inputs // 2, (num_inputs + 1) // 2)):
            if size == 2:
                insts[idx] = BenesNetwork2x2()
            elif size == 3:
                insts[idx] = BenesNetwork3x3()
            else:
                insts[idx] = BenesNetwork(size)
        (self.top, self.bot) = insts

    def reset(self):
        super().reset()
        self.top.reset()
        self.bot.reset()

    def set_switches(self, ord_in, ord_out):
        super().set_switches(ord_in, ord_out)
        ord_out = self.ord_out

        rev_map = [None] * self.num_inputs
        for (idx, val) in enumerate(ord_out):
            rev_map[val] = idx
        is_odd = self.num_inputs % 2 == 1
        iolen = self.num_inputs // 2
        sw_i = [None] * iolen
        sw_o = [None] * (iolen - 1 + is_odd)

        # queue: [ (input_idx, is_top) ]
        invals = set(range(0, self.num_inputs))
        if is_odd:
            invals.discard(self.num_inputs - 1)
            invals.discard(ord_out[-1])
            queue = [ (ord_out[-1], False), (self.num_inputs - 1, False) ]
        else:
            invals.discard(ord_out[-1])
            invals.discard(ord_out[-2])
            queue = [ (ord_out[-1], False), (ord_out[-2], True) ]
        while queue or invals:
            if not queue:
                # if queue is empty, arbitrarily choose an untouched input
                nextval = invals.pop()
                queue.append((nextval, nextval % 2 == 0))
            (idx_i, top_i) = queue.pop()
            swnum_i = idx_i // 2

            # figure out the value for the switch on the input side; set if necessary
            swval_i = (idx_i % 2 == 0) ^ top_i
            if swnum_i >= iolen:
                assert is_odd and swnum_i == iolen, "error: swnum_i overrun"
                assert swval_i == True, "error: tried to set final input switch"
                continue
            if sw_i[swnum_i] is not None:
                assert sw_i[swnum_i] == swval_i, "set_switches error: swval_i"
                continue
            sw_i[swnum_i] = swval_i

            # propagate the constraint for the other side of the switch, if necessary
            nidx_i = idx_i ^ 1
            assert nidx_i < self.num_inputs, "nidx_i out of range"
            invals.discard(nidx_i)
            idx_o = rev_map[nidx_i]
            swnum_o = idx_o // 2
            top_o = not top_i
            swval_o = (idx_o % 2 == 0) ^ top_o

            # figure out the value of the switch on the output side; set if necessary
            if swnum_o == iolen - 1 + is_odd:
                assert swval_o == False ^ is_odd, "error: tried to set final output switch"
                continue
            if sw_o[swnum_o] is not None:
                assert sw_o[swnum_o] == swval_o, "set_switches error: swval_o"
                continue
            sw_o[swnum_o] = swval_o

            # enqueue the new constraint
            nidx_o = idx_o ^ 1
            assert nidx_o < self.num_inputs, "nidx_o out of range"
            if ord_out[nidx_o] in invals:
                queue.append((ord_out[nidx_o], not top_o))
                invals.discard(ord_out[nidx_o])

        self.sw = sw_i + sw_o
        # make sure we're done
        assert None not in self.sw

        top_i = [None] * self.top.num_inputs
        top_o = list(top_i)
        bot_i = [None] * self.bot.num_inputs
        bot_o = list(bot_i)

        if is_odd:
            bot_i[-1] = self.num_inputs - 1
            bot_o[-1] = ord_out[-1]

        sw_o += [False]
        for idx in range(0, iolen):
            (i0, i1) = (2 * idx, 2 * idx + 1)
            (o0, o1) = (ord_out[i0], ord_out[i1])
            (top_i[idx], bot_i[idx]) = (i1, i0) if sw_i[idx] else (i0, i1)
            (top_o[idx], bot_o[idx]) = (o1, o0) if sw_o[idx] else (o0, o1)

        assert sorted(top_i) == sorted(top_o)
        assert sorted(bot_i) == sorted(bot_o)

        self.top.set_switches(top_i, top_o)
        self.bot.set_switches(bot_i, bot_o)

    def route(self, inputs):
        r_exp = super().route(inputs)

        top_i = [None] * self.top.num_inputs
        bot_i = [None] * self.bot.num_inputs

        iolen = self.num_inputs // 2
        is_odd = self.num_inputs % 2 == 1
        sw = self.sw + [False]
        sw_o = sw[iolen:]
        sw_i = sw[:iolen]
        assert sw == sw_i + sw_o

        if is_odd:
            bot_i[-1] = inputs[-1]
        for idx in range(0, iolen):
            (i0, i1) = (inputs[2 * idx], inputs[2 * idx + 1])
            (top_i[idx], bot_i[idx]) = (i1, i0) if sw_i[idx] else (i0, i1)
        assert None not in top_i and None not in bot_i

        top_o = self.top.route(top_i)
        bot_o = self.bot.route(bot_i)

        r = [None] * self.num_inputs
        if is_odd:
            r[-1] = bot_o[-1]
        for idx in range(0, iolen):
            (o0, o1) = (top_o[idx], bot_o[idx])
            (r[2 * idx], r[2 * idx + 1]) = (o1, o0) if sw_o[idx] else (o0, o1)
        assert None not in r
        assert r == r_exp
        return r

if __name__ == "__main__":
    import random
    for _ in range(0, 32):
        length = random.randint(16, 1024)
        foo = list(range(16, 16 + length))
        bar = list(foo)
        quux = BenesNetwork(length)
        for _ in range(0, 32):
            random.shuffle(foo)
            random.shuffle(bar)
            quux.reset()
            quux.set_switches(foo, bar)

            baz = quux.route(foo)
            assert baz == bar

File: benes/benes.hs
{-
 - arbitrary-sized Waksman network impl
 -}

import qualified Control.Monad.ST as CmST
import qualified Data.Array.IO as DaIO
import qualified Data.Array.ST as DaST
import Data.Bits (xor)
import qualified Data.IntSet as DISet
import qualified Data.STRef as DStR

import Control.Exception (assert)
import Control.Monad (liftM, forM_, when)
import qualified Data.List as DLs
import qualified Data.Map.Strict as DM
import Data.Maybe (fromJust)
import qualified System.Environment as Env
import qualified System.Random as Rnd

import qualified Test.QuickCheck as QC

-- precondition: ins is sorted, outs is a permutation of ins
-- returns the indices
toPermOrder :: [Int] -> [Int] -> [Int]
toPermOrder ins outs = res
  where
    iomap = DM.fromDistinctAscList $ zip ins [0..]
    res = map ((DM.!) iomap) outs

-- given a set of switches, split input values into top/bottom lists
benesTopBottom :: [Bool] -> [Int] -> ([Int], [Int])
benesTopBottom sws vals = (reverse i, reverse o)
  where
    bTBHelp (top, bot) [] [] = (top, bot)
    bTBHelp (top, bot) [] (v:[]) = (top, v : bot)
    bTBHelp (top, bot) [] (v:w:[]) = (v : top, w : bot)
    bTBHelp (top, bot) (s:sw) (v:w:xs) =
        let (tt, bb) = if s then (w : top, v : bot) else (v : top, w : bot)
        in bTBHelp (tt, bb) sw xs
    (i, o) = bTBHelp ([], []) sws vals

-- precondition: outp is permutation of inp, inp is sorted
benesRoute :: [Int] -> [Int] -> ([Bool], [Bool])
benesRoute inp outp = CmST.runST $ do
    -- misc values
    let num_inputs = length inp
        is_odd = num_inputs `mod` 2 == 1
        is_odd_ = if is_odd then 1 else 0
        iolen = num_inputs `div` 2
        swnum_o_max = iolen - 1 + is_odd_

    -- input and output switch settings
    sw_i <- DaST.newArray (0, iolen - 1) Nothing :: CmST.ST s (DaST.STArray s Int (Maybe Bool))
    sw_o <- DaST.newArray (0, iolen - 2 + is_odd_) Nothing :: CmST.ST s (DaST.STArray s Int (Maybe Bool))

    -- compute output ordering (maps output idx to input idx)
    let out_ord_ = toPermOrder inp outp
    out_ord <- DaST.newListArray (0, num_inputs - 1) out_ord_ :: CmST.ST s (DaST.STUArray s Int Int)
    -- compute reverse map (maps input idx to output idx)
    rev_map <- DaST.newArray_ (0, num_inputs - 1) :: CmST.ST s (DaST.STUArray s Int Int)
    forM_ (zip out_ord_ [0..]) $ \(val, idx) -> do
        DaST.writeArray rev_map val idx

    -- queue: [ (input_idx, is_top) ]
    -- XXX(rsw) is there a better way to do this? Really, need a STSet type...
    invals <- DStR.newSTRef . DISet.fromDistinctAscList $ take num_inputs [0..]
    queue <- if is_odd
                then do
                    outl1 <- DaST.readArray out_ord (num_inputs - 1)
                    DStR.modifySTRef' invals $ DISet.delete outl1
                    DStR.modifySTRef' invals $ DISet.delete (num_inputs - 1)
                    return [ (num_inputs - 1, False), (outl1, False) ]
                else do
                    outl1 <- DaST.readArray out_ord (num_inputs - 1)
                    DStR.modifySTRef' invals $ DISet.delete outl1
                    outl2 <- DaST.readArray out_ord (num_inputs - 2)
                    DStR.modifySTRef' invals $ DISet.delete outl2
                    return [ (outl2, True), (outl1, False) ]

    -- routing algorithm
    let route_loop = \queue -> do
        -- figure out next queue element to work on, or take one from invals, or we're done
        let qu = DLs.uncons queue
        invals_empty <- liftM DISet.null $ DStR.readSTRef invals
        (elm, queue') <- case (qu, invals_empty) of
                            (Nothing, True) -> return (Nothing, [])
                            (Nothing, False) -> do
                                inval_next <- liftM DISet.findMin $ DStR.readSTRef invals
                                -- XXX need to delete from inval here
                                return (Just (inval_next, inval_next `mod` 2 == 0), [])
                            (Just (h,t), _) -> return (Just h, t)

        when (elm /= Nothing) $ do
            -- input-side switch constraints
            let Just (idx_i, top_i) = elm
                swnum_i = idx_i `div` 2
                swval_i = (idx_i `mod` 2 == 0) /= top_i

            -- input-side switch
            let swnum_i_lim = if swnum_i >= iolen then 0 else swnum_i
            sw_i_val <- DaST.readArray sw_i swnum_i_lim
            if swnum_i >= iolen || sw_i_val /= Nothing
                then route_loop queue'  -- next iteration
                else do
                    DaST.writeArray sw_i swnum_i $! Just swval_i

                    -- output-side switch constraints
                    let nidx_i = idx_i `xor` 1
                    DStR.modifySTRef' invals $ DISet.delete nidx_i
                    idx_o <- DaST.readArray rev_map nidx_i
                    let swnum_o = idx_o `div` 2
                        top_o = not top_i
                        swval_o = (idx_o `mod` 2 == 0) /= top_o

                    -- output-side switch
                    let swnum_o_lim = if swnum_o == swnum_o_max then 0 else swnum_o
                    sw_o_val <- DaST.readArray sw_o swnum_o_lim
                    if swnum_o == swnum_o_max || sw_o_val /= Nothing
                        then route_loop queue'
                        else do
                            DaST.writeArray sw_o swnum_o $! Just swval_o

                            -- new constraint?
                            let nidx_o = idx_o `xor` 1
                            nidx_o_in <- DaST.readArray out_ord nidx_o
                            update_queue <- liftM (DISet.member nidx_o_in) $ DStR.readSTRef invals
                            if not update_queue
                                then route_loop queue'
                                else do
                                    DStR.modifySTRef' invals $ DISet.delete nidx_o_in
                                    route_loop $ (nidx_o_in, not top_o) : queue'

    -- kick off routing
    route_loop queue
    -- return the result
    sw_i_ls <- map fromJust <$> DaST.getElems sw_i
    sw_o_ls <- map fromJust <$> DaST.getElems sw_o
    return (sw_i_ls, sw_o_ls)

benesRoute3 :: [Int] -> [Int] -> [Bool]
benesRoute3 inp outp = [sw_i, sw_m, sw_o]
  where
    [ord0, ord1, ord2] = toPermOrder inp outp
    (sw_m, sw_i, sw_o) = if ord2 /= 2
                            then (True, ord2 == 0, ord1 /= 2)
                            else (False, False, ord0 == 1)

benesBuild :: String -> [Int] -> [Int] -> IO ()
benesBuild name inp outp = do
    let num_inputs = length inp
        is_odd = num_inputs `mod` 2 == 1
        iolen = num_inputs `div` 2
        (sw_i, sw_o) = benesRoute inp outp

    flip mapM (zip [0..] sw_i) $ \(idx, swval) -> do
        let nameI0 = name ++ "_wI_" ++ show (2 * idx)
            nameI1 = name ++ "_wI_" ++ show (2 * idx + 1)
            nameOT = name ++ "_top_wI_" ++ show idx
            nameOB = name ++ "_bot_wI_" ++ show idx
            nbool = name ++ "_swI_" ++ show idx
        print $ nameOT ++ " = if " ++ nbool ++ " then " ++ nameI1 ++ " else " ++ nameI0
        print $ nameOB ++ " = if " ++ nbool ++ " then " ++ nameI0 ++ " else " ++ nameI1
        print $ nbool ++ " = " ++ show swval
    if is_odd
        then do
            let nameI = name ++ "_wI_" ++ show (num_inputs - 1)
                nameO = name ++ "_bot_wI_" ++ show (length sw_i)
            print $ nameO ++ " = " ++ nameI
        else return ()

    flip mapM (zip [0..] sw_o) $ \(idx, swval) -> do
        let nameIT = name ++ "_top_wO_" ++ show idx
            nameIB = name ++ "_bot_wO_" ++ show idx
            nameO0 = name ++ "_wO_" ++ show (2 * idx)
            nameO1 = name ++ "_wO_" ++ show (2 * idx + 1)
            nbool = name ++ "_swO_" ++ show idx
        print $ nameO0 ++ " = if " ++ nbool ++ " then " ++ nameIB ++ " else " ++ nameIT
        print $ nameO1 ++ " = if " ++ nbool ++ " then " ++ nameIT ++ " else " ++ nameIB
        print $ nbool ++ " = " ++ show swval
    if is_odd
        then do
            let nameI = name ++ "_bot_wO_" ++ show (length sw_o)
                nameO = name ++ "_wO_" ++ show (num_inputs - 1)
            print $ nameO ++ " = " ++ nameI
        else do
            let nameI0 = name ++ "_top_wO_" ++ show (length sw_o)
                nameI1 = name ++ "_bot_wO_" ++ show (length sw_o)
                nameO0 = name ++ "_wO_" ++ show (num_inputs - 2)
                nameO1 = name ++ "_wO_" ++ show (num_inputs - 1)
            print $ nameO0 ++ " = " ++ nameI0
            print $ nameO1 ++ " = " ++ nameI1

    let (it, ib) = benesTopBottom sw_i inp
        (ot, ob) = benesTopBottom sw_o outp
    print (it, ib)
    print (ot, ob)

    return ()

-- *** tests infra ***
-- check that Benes is OK
test_benesRoute :: Int -> QC.Gen Bool
test_benesRoute len = do
    let ivals = take (4 + len `mod` 1024) [0..]
    ovals <- QC.shuffle ivals
    let (sw_i, sw_o) = benesRoute ivals ovals
        (it, ib) = benesTopBottom sw_i ivals
        (ot, ob) = benesTopBottom sw_o ovals
        imatch = it == DLs.sort ot
        omatch = ib == DLs.sort ob
    return $ imatch && omatch

main :: IO ()
main = do
    max_loops <- liftM read (liftM head $ Env.getArgs)
    QC.quickCheck $ QC.withMaxSuccess max_loops test_benesRoute
